NDTensors/src/linearalgebra.jl

```diff
@@ -388,23 +367,120 @@ function LinearAlgebra.eigen(
   return D, V, spec
 end
 
-function LinearAlgebra.qr(T::DenseTensor{ElT,2,IndsT}; kwargs...) where {ElT,IndsT}
-  positive = get(kwargs, :positive, false)
-  # TODO: just call qr on T directly (make sure
-  # that is fast)
-  if positive
-    QM, RM = qr_positive(matrix(T))
-  else
-    QM, RM = qr(matrix(T))
-  end
+function qr(T::DenseTensor{ElT,2,IndsT}; positive=false, kwargs...) where {ElT,IndsT}
+  qxf = positive ? qr_positive : qr
+  return qx(qxf, T; kwargs...)
+end
+function ql(T::DenseTensor{ElT,2,IndsT}; positive=false, kwargs...) where {ElT,IndsT}
+  qxf = positive ? ql_positive : ql
+  return qx(qxf, T; kwargs...)
+end
+
+#
+#  Generic function for qr and ql decomposition of dense matrix.
+#  The X tensor = R or L.
+#
+function qx(qx::Function, T::DenseTensor{ElT,2,IndsT}; kwargs...) where {ElT,IndsT}
+  QM, XM = qx(matrix(T))
   # Make the new indices to go onto Q and R
   q, r = inds(T)
   q = dim(q) < dim(r) ? sim(q) : sim(r)
   Qinds = IndsT((ind(T, 1), q))
-  Rinds = IndsT((q, ind(T, 2)))
-  Q = tensor(Dense(vec(Matrix(QM))), Qinds)
-  R = tensor(Dense(vec(RM)), Rinds)
-  return Q, R
+  Xinds = IndsT((q, ind(T, 2)))
+  Q = tensor(Dense(vec(Matrix(QM))), Qinds) #Q was strided
+  X = tensor(Dense(vec(XM)), Xinds)
+  return Q, X
+end
+
+#
+# Just flip signs between Q and R to get all the diagonals of R >=0.
+# For rectangular M the indexing for "diagonal" is non-trivial.
+#
+"""
+    qr_positive(M::AbstractMatrix)
+
+Compute the QR decomposition of a matrix M
+such that the diagonal elements of R are
+non-negative. Such a QR decomposition of a
+matrix is unique. Returns a tuple (Q,R).
+"""
+function qr_positive(M::AbstractMatrix)
+  sparseQ, R = qr(M)
+  Q = convert(Matrix, sparseQ)
+  nc = size(Q, 2)
+  for c in 1:nc
+    if real(R[c, c]) < 0.0
+      R[c, c:end] *= -1 #only fip non-zero portion of the row.
+      Q[:, c] *= -1
+    end
+  end
+  return (Q, R)
+end
+
+"""
+    ql_positive(M::AbstractMatrix)
+
+Compute the QL decomposition of a matrix M
+such that the diagonal elements of L are
+non-negative. Such a QL decomposition of a
+matrix is unique. Returns a tuple (Q,L).
+"""
+function ql_positive(M::AbstractMatrix)
+  sparseQ, L = ql(M)
+  Q = convert(Matrix, sparseQ)
+  nr, nc = size(L)
+  dc = nc > nr ? nc - nr : 0 #diag is shifted over by dc if nc>nr
+  for c in 1:(nc - dc)
+    if c <= nr && real(L[c, c + dc]) < 0.0
+      L[c, 1:(c + dc)] *= -1 #only fip non-zero portion of the column.
+      Q[:, c] *= -1
+    end
+  end
+  return (Q, L)
+end
+
+#
+#  Lapack replaces A with Q & R carefully packed together.  So here we just copy a
+#  before letting lapack overwirte it. 
+#
+function ql(A::AbstractMatrix{T}; kwargs...) where {T}
```

Comment:
What about using https://github.com/JuliaLinearAlgebra/MatrixFactorizations.jl for this?

Reply:
Yes I had a look at that.  in ql.jl line 182 they actually the (almost) identical code:
function _ql(layout, axes, A, args...; kwds...)
    require_one_based_indexing(A)
    AA = similar(A, _qleltype(eltype(A)), size(A))
    copyto!(AA, A)
    return ql!(AA, args...; kwds...)
end

I suspect the reason it is identical is becuase they looked at qr(A::AbstractMatrix) inside LinearAlgebra.  In LinearAlgebra the subsequent call to qr!(A::AbstractMatrix) is just a call to lapack like we have.  Whereas the  https://github.com/JuliaLinearAlgebra/MatrixFactorizations.jl  people seem to be doing something more complicated.  

If you are trying to avoid the lapack call I don't see how unless there some algebraic relation between qr and ql that I am not aware of.

Reply:
I just would prefer to use that package so we don't have to support our own wrapper (I assumed they would both be using LAPACK).

Reply:
It is giving the wrong answer for rectungular input with N_rows>N_columns.  I also notice they have no unit test for ql, rq?!?
This has prompted me to beef up the unit tests at the NDTensors level to catch this sort of thing.

Reply:
I see, that's too bad. Would you mind raising an issue in that repository about that? It would be nice to just use that package and have one community-wide version of `ql` but I guess we will need to support our own for now. Really it should be in Julia's standard LinearAlgebra library but I guess no one bothered to make a high level wrapper for it.

Reply:
Sure.  I can also save a git-stash of this change so we can resurect later.

Reply:
[Done](https://github.com/JuliaLinearAlgebra/MatrixFactorizations.jl/issues/43)

Reply:
Great, thanks! Yes hopefully that issue is easy to fix so we can just use their wrapper (though of course that doesn't need to block this PR).